gerritbot-bot/src/lib.rs

```diff
@@ -202,44 +197,449 @@ impl Bot {
+        })
+    }
+
+    pub fn build(
+        self,
+        gerrit_command_runner: gerrit::CommandRunner,
+        spark_client: spark::Client,
+    ) -> Bot {
+        let Self {
+            format_script,
+            msg_cache_parameters,
+            state,
+        } = self;
+
+        Bot {
+            gerrit_command_runner,
+            spark_client,
+            msg_cache: msg_cache_parameters.map(
+                |MsgCacheParameters {
+                     capacity,
+                     expiration,
+                 }| {
+                    LruCache::<MsgCacheLine, ()>::with_expiry_duration_and_capacity(
+                        expiration, capacity,
+                    )
+                },
+            ),
+            format_script,
+            state,
+        }
+    }
+}
+
+fn spark_message_to_action(message: spark::Message) -> Action {
+    lazy_static! {
+        static ref FILTER_REGEX: Regex = Regex::new(r"(?i)^filter (.*)$").unwrap();
+    };
+
+    let sender_email = message.person_email;
+    let sender_id = message.person_id;
+    match &message.text.trim().to_lowercase()[..] {
+        "enable" => Action::Enable(sender_id, sender_email),
+        "disable" => Action::Disable(sender_id, sender_email),
+        "status" => Action::Status(sender_id),
+        "help" => Action::Help(sender_id),
+        "filter" => Action::FilterStatus(sender_id),
+        "filter enable" => Action::FilterEnable(sender_id),
+        "filter disable" => Action::FilterDisable(sender_id),
+        _ => FILTER_REGEX
+            .captures(&message.text.trim()[..])
+            .and_then(|cap| cap.get(1))
+            .map(|m| Action::FilterAdd(sender_id.clone(), m.as_str().to_string()))
+            .unwrap_or_else(|| Action::Unknown(sender_id.clone())),
+    }
+}
+
+/// Transform a gerrit event into a bot action.
+pub fn gerrit_event_to_action(event: gerrit::Event) -> Option<Action> {
+    if event.event_type == gerrit::EventType::CommentAdded && event.approvals.is_some() {
+        Some(Action::UpdateApprovals(Box::new(event)))
+    } else if event.event_type == gerrit::EventType::ReviewerAdded {
+        Some(Action::ReviewerAdded(Box::new(event)))
+    } else {
+        None
+    }
+}
+
+impl Bot {
+    pub fn run(
+        self,
+        // TODO: gerrit event stream probably shouldn't produce errors
+        gerrit_events: impl Stream<Item = gerrit::Event, Error = String> + Send,
+        spark_messages: impl Stream<Item = spark::Message, Error = ()> + Send,
+    ) -> impl Future<Item = (), Error = ()> {
+        let _ = &self.gerrit_command_runner;
+        let spark_client = self.spark_client.clone();
+        let gerrit_actions = gerrit_events
+            .filter_map(gerrit_event_to_action)
+            .map_err(|e| error!("failed to receive gerrit event: {}", e));
+        let spark_actions = spark_messages.map(spark_message_to_action);
+        let bot_for_action = std::sync::Arc::new(std::sync::Mutex::new(self));
```

Comment:
This completely brakes the paradigm the bot was using before. No mutexes and locking + a single threaded event loop. This was based on so called elm-architecture (https://guide.elm-lang.org/architecture/). I appreciate all other refactoring, but I am not sure I am willing to change that.

Reply:
There's still a single event loop.  However, `tokio` now uses a multi-threaded runtime by default which executes `Future`s on a thread pool and therefore requires them to be `Send`.  I think it might be possible to avoid `Arc` and `Mutex` here and still have the resulting future `Send` since `update` and `task` don't return futures themselves.